Add RFPR blog post

collinvandyck opened the pull request:
This blog post takes what had originally been a PDF that we have been sharing around periodically and turns it into a URL that can be more easily shared, and updated over time.

The content is basically the same as that of the PDF if you squint, but it goes into more detail than the PDF did.

collinvandyck committed:
Add jekyll-target-blank jekyll plugin.

This automatically sets target=_blank for any external link, like those to
lazygit.

collinvandyck committed:
Add RFPR blog post and lazygit screenshot.

collinvandyck committed:
Address spelling issues.

collinvandyck committed:
Add section addressing commit mistakes.

lmarburger commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,347 @@
+I use this constantly and so my feature branch is littered with `wip` commits. These are my
+breadcrumbs that enable to me to "checkpoint" my progress in a way that doesn't take me out of the
+flow. They also let me back out more granular changes -- maybe I wanted to see if a particular change
+made sense. I `ci` before I start, and if I hate it, it's easy for me to throw away that change
+without risk of throwing away what came before it.
```
This is a cool idea. I end up relying a lot on undo in my editor which isn't as reliable. I need to get better at using git to checkpoint my work.

collinvandyck replied:
I think that it shines in particular when the work you're doing is outside of the editor itself (e.g. adding a migration, doing some codegen of some kind, etc). This also makes me realize that I really don't rely on neovim's undo much at all, and I should probably lean into that at some point. Specifically, what does it look like to undo changes across multiple buffers in a coherent way -- no idea haha.

lmarburger commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,347 @@
+
+I would highly recommend *not* rebasing or rewriting commits after you start getting feedback. The
+reason for this is that reordering the commits or amending them, ends up creating new commits under
+the hood (even if `lazygit` made it super easy to do so). Because reviewers will leave feedback on
+each commit during the initial review, if you rewrite that commit as a followup and force push, the
+feedback they left becomes orphaned and out of context because that commit no longer exists.
```
This is interesting. When I'm reviewing a PR that has good commits, I have another tab open with the PR's files tab open where I leave comments in-line. That's mostly because the way GitHub handles comments on a commit is really awkward and, frankly, I don't trust it. Maybe that's because I've been burned by comments I've left on commits that become orphaned and are lost.

collinvandyck replied:
Man, we are all GitHub prisoners 😂 I think phabricator handles this particular scenario with more grace, but I only used it once, and that was in 2012 at Boundary. There was a lot at that time that it did *not* handle with grace, and I was relieved when GitHub PRs became the standard, even with its warts.

lmarburger commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,347 @@
+
+1. Commits should be limited in scope.
+2. Do not mix high value and low value concerns in the same commit.
+3. Commit message titles should be a sentence that describes an action.
+4. Commit messages should end with a period.
```
I don't know that point 4 adds anything that isn't covered by 3. Point 3 captures the spirit by saying commit messages should be sentences and your examples below drive that home by including punctuation.

What's the intent? If it's to make it clear that commit messages should be descriptive and not lazy shorthand, what about saying in 3 that a commit title should be a "full sentence"?

Personally, my commit message are more like PR titles than full sentences with punctuation. (I don't know the name for this style. It's somewhere between a newspaper article title without title case and a sentence.) I'm not opposed to using a period in commit message, but I don't use them. If I was on your team I'd have no problem falling in line. 😄

collinvandyck replied:
Yeah, good point. I don't think point 4 is worth keeping. I yoinked it out and added your suggestion about "full sentence".

lmarburger reviewed (approved):
I love seeing this getting a proper blog post so I can more easily share it everywhere I go. ❤️ 

It's great to see the method has evolved to lean on `lazygit` instead of spending a lot of time on git minutia. I think it's a good move. 💯 If you know of a quick, simple `lazygit` demo that shows someone building incremental commits, it might be useful for someone brand new to see. The tutorials are great but the ones I've seen tend to be a whirlwind tour of a ton of features.

I don't have any feedback on the overall RFPR method. It's been honed at this point and I'm bought-in. 🚀

lmarburger commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,347 @@
+each commit during the initial review, if you rewrite that commit as a followup and force push, the
+feedback they left becomes orphaned and out of context because that commit no longer exists.
+
+For that reason, it's better to push new commits when addressing feedback, even if those changes
+would normally be structured differently if you were creating the RFPR from scratch.
```
Something I just thought of reading this section. What's your stance on rebase+merge or squash-and-merge of PRs? Do you merge to transfer the commit history to `main` or do you squash it all together?

I'm a strong supporter of the squash-and-merge approach. The commits I build are meant for reviewers and I don't find them useful outside the context of a PR. When I `git blame` a line, I'm perfectly happy to open my browser and read the PR that created it. Not rewriting the commit history after people start reviewing means the commits at the end of a review may not be pristine.

Anyway, just a thought I had. It may be worthwhile to at least touch on the tradeoffs if you don't have a strong opinion about which method is "right".

collinvandyck replied:
Thanks for bringing this up. I added a section at the end for this. The TLDR is that I don't feel strongly about it either way but I explained how I typically think about it. LMK what you think.

collinvandyck commented:
>If you know of a quick, simple lazygit demo that shows someone building incremental commits, it might be useful for someone brand new to see. The tutorials are great but the ones I've seen tend to be a whirlwind tour of a ton of features.

Yeah I agree. I went back in my archives and I found the video presentation of these ideas that I did at ngrok, but the "live demo" part uses `tig` instead of `lazygit`. I was kind of hoping that I could leverage that, but I agree with your opinion that `lazygit` is probably the right thing to focus on going forward.

Maybe that's something one of us could do at some point in the future.

collinvandyck committed:
Fix some grammar issues when talking about reordering or amending commits.

collinvandyck committed:
Remove low-value point about commits ending in a sentence.

collinvandyck committed:
Add a section on merge commit vs squash merge.

mheffner commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,372 @@
+can be exhausting and therefore understandable that we would feel the impulse to create the PR as
+quickly as possible to get this over the finish line. I hope you'll agree that by the end of this
+post that spending a bit more effort to create a better PR is worth it, and is probably not as much
+effort as one might think.
```
You allude to it a bit later, but I feel the value of a well documented PR lives on past just the initial review too. If this is a service running in production, someone else on-call that didn't review the PR, may need to refer back to it when there's an outage and they are trying to understand what happened. If this is an OSS project, people may want to review the PRs that land in the next release to better understand how the project is changing. I guess to say, the time spent in making the PR clear to understand can pay out dividends many times over.

mheffner commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,372 @@
+
+> Time spent creating a better PR costs the business less to merge it.
+
+# RFPR Overview
```
Expand acronym in title?

mheffner commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,372 @@
+# RFPR Overview
+
+Reviewer First Pull Requests (RFPRs) are a prescriptive approach to building better PRs that are
+cheaper to review. At Mesosphere I worked on a team under [Kevin
```
Maybe use a different phrasing than "cheaper"?  Eg: "easier", "that require less cognitive load", ?

mheffner commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,372 @@
+
+Reviewer First Pull Requests (RFPRs) are a prescriptive approach to building better PRs that are
+cheaper to review. At Mesosphere I worked on a team under [Kevin
+Klues](https://www.linkedin.com/in/klueska). After creating a PR which was hard to review he
```
"One time after I published a large PR that was hard to review, Kevin"...

mheffner commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,372 @@
+
+The PR description sets the stage for the reviewer. Since they are going to review the PR
+commit-by-commit, the description should give them as much context as possible before they start.
+Assume that the reviewer does not have the context that you do.
```
I have to admit, I honestly don't usually review PRs commit-by-commit, for many of the reasons you document here -- people haven't practiced this discipline. I wonder how many people actually know how to review PRs commit-by-commit in Github and other tools. It could be worth demonstrating how to do this since most folks likely just pull up the entire changeset at once.

collinvandyck replied:
That's a great callout. I'll add an appendix for this.

lmarburger replied:
I explicitly call out at the top of the PR that I suggest reviewers review commit-by-commit. I'm a broken record at this point. I think the default mode is to just slog through the full diff.

mheffner commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,372 @@
+forgotten to address during development. By writing the description *for the reviewer* this causes
+you to think about the things that the reviewer would probably have questions about. Anything you
+catch at this phase is a major win, as it is a short amount of time compared to the revisions to the
+PR which would otherwise follow.
```
Would it be worth adding some mention of when it may be a better idea to split a large PR not by commits, but actually into separate PRs themselves? How would you define the demarcations for separate PRs and would it be based on granular deployment of code to a production environment?

collinvandyck replied:
I feel like this is probably pretty difficult to talk about without adding a lot more content. I think you're right that a lot of it has to do with how a team does deployments, and whether or not you're incrementally deploying something. It seems related, but somewhat orthogonal (ten dollar word) to RFPRs.

lmarburger replied:
That's a good point. My experience with RFPRs has been that the process of building these commits makes it obvious when I have substantial unrelated work. At that point it's easy enough to move the commits to a new branch and PR.

mheffner commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,372 @@
+- (nothing)
+
+Imagine that you are trying to find a particular PR on GitHub during an incident. The PR title
+should be descriptive enough so that it can be easily identified.
```
Worth a mention or sharing your thoughts on [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)? I don't prescribe to it, but many people do so it may come up in comments.

collinvandyck replied:
I don't actually have many thoughts on it. I've only used it at one company, and the adoption of it was pretty mixed, and I'm not sure I've ever gotten a lot of value out of it, so I don't think I could write anything meaningful about it.

mheffner commented on _posts/2025-02-13-rfprs.md:
```diff
@@ -0,0 +1,372 @@
+
+These are the kinds of PRs I'm thankful to get notified about. I might spend on average 15-30m extra
+to create PRs using these guidelines, but it saves many hours across the reviewers assigned to it,
+and the quality of feedback you get on your RFPR will invariably be higher.
```
Any PRs out there on public repos that could be linked to as examples?

collinvandyck replied:
asked in slack

mheffner reviewed (commented):
Looks good, nice job putting this together! I left a few high-level thoughts/comments, feel free to take or leave them.

I was able to read it easily and understand it well, but there a few instances where a sentence or two seemed long. I have found in my own writing that dumping content I've written in to the LLMs, ChatGPT seems decent, and asking for grammar feedback has been useful. I pick and chose what edits I prefer over my own.

Excited to see this go live!

collinvandyck committed:
Address feedback from Mike.

This also includes a section at the end on how to use GH's keyboard shortcuts to
make commit-order reviewing a little easier.

collinvandyck committed:
Add table of contents.

There's probably a better way to do this, but the gem I tried didn't seem to
work.
